### ISSUE TITLE
What is the Process Or Functionality For Stream Management In layman's term

### ISSUE THREAD

**Issue by coderRauthan** [processone/ejabberd#3086]
Hi Friends I'm Using Ejabberd 19.02.

So the thing is currently there is an issue of Message Drop when user turns off Internet or the network connection got lost.If At that time somebody message that user,those messages will be lost.As presence of that user will be online and those messages will not be stored in spool table.

Scenario - User A is online,User B is Online.They messages each other.Everything works fine.But Suddenly User B internet got disconnected.Now User B is disconnected but his presence is still online.Now the messages that are sent by User A to User B are lost.

Solution - I've Read that Stream Management is created specifically for this task.As its Acknowledges from both server and client side.So I've Implemented it with the help of https://docs.ejabberd.im/admin/configuration/#mod-stream-mgmt .I've set my setting as :  

max_ack_queue                --->  15000
max_resume_timeout:	--->	0
resend_on_timeout		---> true(I know for MUC it should be if_offline,currently true for testing)
resume_timeout		---> 0
ack_timeout			--->	1

I've enabled Stream Management From Both Server(Ejabberd 19.02) and Client(Smack 4.2) Side.

But Currently the issue is,If User B is Offline and User A sends messages,Then:-

The first Message is getting lost.The process will be -  
User B is offline but presence will be shown as online.User A sends message,server will not get acknowledgement from User B. At timeout the user B will get disconnected and from there on
all messages will be stored in spool table.But In this case the first message will always get lost.

I dont want to use PingManager, as in my opinion it is an expensive solution,And Even with Ping there will be a window in which messages will be lost.

I dont want to increase resume_timeout ,As i  need to create new session if User is disconnected once.

So here are my 2 questions:-

1) How does Stream Management work end to end in layman's term,I am sure My understanding of it is not complete.Like I have many questions regarding it:-
a) Where are the messages stored when user is disconnected,but his presence is online.On Connection resumption how does he get message when the messages are not stored in spool table.
b) What will happen if user is disconnected,But his presence is online.If Stream Times out,where will the message go.Will it be stored in spool table or not.How will user get his message when he connects next time.

2) How To Solve This issue,My first message will always be lost(which makes sense to me,as it will not be able to send acknowledgement back,and then server will disconnect the user,and next messages will be stored offline but that message will be lost).

I'm sorry if something sounds stupid,but there are no resources in which end to end Stream Management process is explain clearly.It will help me a lot if Somebody explain Stream Management theoretically.And how to solve this issue.

Thanks,
Best Regards

**Comment by weiss** [comment]
The [stream management spec][1] and [an old-ish FAQ entry][2] might clarify some things.  Note that the FAQ article was written before [server-side message archiving (MAM)][3] was a widespread thing, which would be today's main mechanism to avoid message loss (and to implement multi-device support).  With MAM support in place, stream management mostly boils down to being an optimization.

> I dont want to increase resume_timeout ,As i need to create new session if User is disconnected once.

Where does this need arise from?  The main point of stream management is allowing the client to quickly [resume the old session][4] and thereby receive any missed stanzas after some short network outage, rather than creating a new session.

Anyway, you're not describing an ejabberd bug, so if you have further questions, please ask them in some support venue such as the [mailing list][5] or the chat room (`ejabberd@conference.process-one.net`).

[1]: https://xmpp.org/extensions/xep-0198.html
[2]: https://www.ejabberd.im/faq/tcp
[3]: https://xmpp.org/extensions/xep-0313.html
[4]: https://xmpp.org/extensions/xep-0198.html#resumption
[5]: http://lists.jabber.ru/mailman/listinfo/ejabberd

**Comment by licaon-kter** [comment]
Messages don't get lost if you enable MAM and Offline.

Use the example config first ( https://github.com/processone/ejabberd/blob/master/ejabberd.yml.example ), then modify it to your case.

**Comment by coderRauthan** [comment]
@weiss Thanks,will look into this.

@licaon-kter Thanks, for the help.But I've already enable mod_mam and mod_offline.And all chats are stored in our database.

Are you saying that if Stream Times out and then user gets disconnect from server.Then those messages will be retrieved from archieve table? Is this the standard and recommended way?

Thanks,
Best Regards

**Comment by licaon-kter** [comment]
@coderRauthan maybe there is something else wrong


Can you test with https://github.com/iNPUTmice/caas ?

**Comment by coderRauthan** [comment]
@licaon-kter got it.You were right eventually,there is a configuration issue.

I've configured my mod_mam as:-
 assume_mam_usage: true

So,according to docs if Stream Closes(TimeOut) then the messages which were sent before timeout will not be stored in offline table(spool),It is assumed that those messages will be retrieved from the archive. 

Is there any configuration,in which these buffer messages can be stored in Offline Table? Or how to deal with it without retrieving messages from archieve table.

Many many thanks,Now i atleast understand that what is going on.

**Comment by weiss** [comment]
> Is there any configuration,in which these buffer messages can be stored in Offline Table?

```yaml
modules:
  mod_stream_mgmt:
    resend_on_timeout: true
```

Note you will only want to set it this way if you _don't_ support multi-device usage (see the [documentation][1] and the [FAQ article I linked above][2] for details).  For multi-device support, you want MAM.

[1]: https://docs.ejabberd.im/admin/configuration/#mod-stream-mgmt
[2]: https://www.ejabberd.im/faq/tcp

**Comment by licaon-kter** [comment]
@coderRauthan You don't want messages in MAM and/or offline?
```
mod_mam:
    assume_mam_usage: true
    default: always
    compress_xml: true
  mod_muc:
    default_room_options:
      mam: true
  mod_offline:
    access_max_user_messages: max_user_offline_messages
    store_groupchat: true
    use_mam_for_storage: true
```

**Comment by zinid** [comment]
@licaon-kter the indentation is borked :)

**Comment by coderRauthan** [comment]
@weiss and @licaon-kter thanks man,I am able to solve the issue with the help of two of you and in the process I've also learnt how Stream Management works.

What I wanted was that all messages to be stored in archive table and also if any message is not delivered to the receiver(In the case of Stream Connection Time Out then those messages to be stored in spool table untill they are delivered to the receiver).

The Configuration for this is:-
mod_mam:-
          db_type: sql
          assume_mam_usage: false
          default: always
And mod_stream_mgmt:-
       resend_on_timeout: true

Again thanks @weiss and @licaon-kter  I also understand the process Of Stream Management and ejabberd mechanism for mod_stream_mgmt.

I know It wasn't an ejabberd bug,but yet you guys volunteer.
Thanks,
Best Regards
